<a href="https://colab.research.google.com/github/millyazhukovich/machine_learning2024/blob/Zhukovich/%D0%96%D1%83%D0%BA%D0%BE%D0%B2%D0%B8%D1%87__%D0%A0%D0%90%D0%97_3%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1. **Для данных по качеству вин постройте многоклассовую классификацию. Загрузка
данных по качеству вин была рассмотрена на одном из занятий ранее.**

In [33]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import LabelEncoder
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline

In [2]:
data = load_wine()
X = data.data
y = data.target

In [3]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [4]:
model = LogisticRegression(multi_class='multinomial', solver='lbfgs', max_iter=200)
model.fit(X_train, y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(max_iter=200, multi_class='multinomial')

In [5]:
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)


In [7]:
print("Точность модели (Accuracy):", accuracy)
print("\nМатрица ошибок (Confusion Matrix):\n", conf_matrix)
print("\nОтчет о классификации (Classification Report):\n", class_report)
print("Целевые метки для индексов 10, 80, 140:", data.target[[10, 80, 140]])
print("Классы вин:", list(data.target_names))

Точность модели (Accuracy): 0.9722222222222222

Матрица ошибок (Confusion Matrix):
 [[13  1  0]
 [ 0 14  0]
 [ 0  0  8]]

Отчет о классификации (Classification Report):
               precision    recall  f1-score   support

           0       1.00      0.93      0.96        14
           1       0.93      1.00      0.97        14
           2       1.00      1.00      1.00         8

    accuracy                           0.97        36
   macro avg       0.98      0.98      0.98        36
weighted avg       0.97      0.97      0.97        36

Целевые метки для индексов 10, 80, 140: [0 1 2]
Классы вин: ['class_0', 'class_1', 'class_2']


**2. Выберите любой набор данных на kaggle.com, пригодный для проведения
многоклассовой классификации (вспомните, чем классификация отличается от
регрессии). Выберу набор данных test.tsv.Датасет содержит информацию о товарах, выставленных на продажу на платформе Mercari, с различными признаками**

In [13]:
df = pd.read_csv('test.tsv', sep='\t')
print("Пример данных:")
print(display(df.head()))


Пример данных:


test_id                                      name  item_condition_id  \
0        0  Breast cancer "I fight like a girl" ring                  1   
1        1  25 pcs NEW 7.5"x12" Kraft Bubble Mailers                  1   
2        2                                 Coach bag                  1   
3        3                             Floral Kimono                  2   
4        4                          Life after Death                  3   

                                    category_name brand_name  shipping  \
0                             Women/Jewelry/Rings        NaN         1   
1         Other/Office supplies/Shipping Supplies        NaN         1   
2  Vintage & Collectibles/Bags and Purses/Handbag      Coach         1   
3                         Women/Sweaters/Cardigan        NaN         0   
4             Other/Books/Religion & Spirituality        NaN         1   

                                    item_description  
0                                             Size 7  
1  25 pcs NEW 7.5"x12" Kraft Bubble Mailers Lined...  
2  Brand new coach bag. Bought for [rm] at a Coac...  
3  -floral kimono -never worn -lightweight and pe...  
4  Rediscovering life after the loss of a loved o...

None


**3. Выполните его очистку: удалите пропуски (dropna() в pandas), закодируйте
строковые данные (можно LabelEncoder() из Scikit-Learn). Пример очистки в коде
example.py.**

Пример:
1. Удаление строк с пропусками

df_dropped = df.dropna()

print("\nDataFrame после удаления строк с пропусками:")

print(df_dropped)

2. Кодирование строковых значений с помощью LabelEncoder

label_encoder = LabelEncoder()

df_filled['City'] = label_encoder.fit_transform(df_filled['City'])

print("\nDataFrame после кодирования строковых значений:")

print(df_filled)
В моем случае:
Столбец name, category_name, brand_name, и item_description были преобразованы в числовые значения.
Столбец shipping остался числовым, так как он уже содержал 0 и 1.

In [19]:
# Удаление строк с пропусками
df_cleaned = df.dropna()

# Инициализация кодировщика
label_encoder = LabelEncoder()

# Применение LabelEncoder к строковым столбцам
for column in df_cleaned.select_dtypes(include=['object']).columns:
    df_cleaned[column] = label_encoder.fit_transform(df_cleaned[column])
df_cleaned.head()

test_id    name  item_condition_id  category_name  brand_name  shipping  \
2         2   73128                  1            591         813         1   
6         6  313273                  3            738        3649         1   
8         8   79238                  3             11        2133         0   
12       12  284976                  3            724         882         0   
13       13  163852                  3            694        2114         0   

    item_description  
2              67314  
6             307657  
8              64462  
12             14766  
13            283911

**4. Постройте модель многоклассовой логистической регрессии и оцените ее качество.**

In [25]:
X = df.drop(columns=['test_id', 'item_condition_id'])
y = df['item_condition_id']

In [34]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Создание пайплайна
preprocessor = ColumnTransformer(
    transformers=[
        ('text', TfidfVectorizer(), 'item_description'),  # Преобразуем текстовое поле с помощью TF-IDF
        ('cat', OneHotEncoder(handle_unknown='ignore'), ['category_name']),  # Преобразуем категориальные признаки
        ('num', SimpleImputer(strategy='mean'), ['shipping'])  # Числовые пропущенные значения
    ])

model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', LogisticRegression(multi_class='multinomial', solver='lbfgs', max_iter=200))
])

# Обучение модели
model.fit(X_train, y_train)

# Предсказания на тестовой выборке
y_pred = model.predict(X_test)

# Оценка качества модели
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

# Вывод результатов
print("Точность модели (Accuracy):", accuracy)
print("\nМатрица ошибок (Confusion Matrix):\n", conf_matrix)
print("\nОтчет о классификации (Classification Report):\n", class_report)

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:1247: FutureWarning: 'multi_class' was deprecated in version 1.5 and will be removed in 1.7. From then on, it will always use 'multinomial'. Leave it to its default value to avoid this warning.
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


Точность модели (Accuracy): 0.7086533731885429

Матрица ошибок (Confusion Matrix):
 [[24300  2552  1596     9     0]
 [ 3615 10875  7505    14     2]
 [ 1531  4298 20636   193     7]
 [   91   109  1429   338    13]
 [    6    12    78    40    38]]

Отчет о классификации (Classification Report):
               precision    recall  f1-score   support

           1       0.82      0.85      0.84     28457
           2       0.61      0.49      0.55     22011
           3       0.66      0.77      0.71     26665
           4       0.57      0.17      0.26      1980
           5       0.63      0.22      0.32       174

    accuracy                           0.71     79287
   macro avg       0.66      0.50      0.54     79287
weighted avg       0.70      0.71      0.70     79287



**5. Сделайте выводы о качестве полученных моделей и причины таких результатов.**

Вывод по первой модели (качество вин).

1. Точность модели (Accuracy):
Точность модели составила 97.2%, что является отличным результатом, особенно для задачи многоклассовой классификации.

2. Матрица ошибок (Confusion Matrix):
В матрице ошибок видно, что все классы классифицируются с очень высокой точностью. Например, для класса 0 точность составила 1.00 (правильные предсказания) с ошибками только по одному экземпляру (13 из 14 предсказаны правильно).
Также видно, что модель не ошибается на классах 1 и 2, что указывает на высокое качество предсказания.

3. Отчет о классификации (Classification Report):
Precision, Recall и F1-score для всех классов близки к 1.00, что означает, что модель очень хорошо справляется с классификацией каждого из классов.
Важно отметить, что Recall и F1-score также высоки, что говорит о том, что модель не только точно предсказывает, но и не пропускает ошибки.
Значение Macro avg и Weighted avg также подтверждают, что модель обрабатывает все классы с одинаково хорошими результатами.

Причины таких результатов:

Логистическая регрессия как модель: Это довольно простая и эффективная модель для многоклассовой классификации, особенно когда данные хорошо линейно разделимы.

Хорошая репрезентация классов в данных: Проблемы с классификацией обычно возникают, когда классы плохо представлены или перекрываются, но в данном случае классы, вероятно, хорошо разделимы.

Параметры модели: Модель использовала многоклассовую логистическую регрессию с оптимизатором lbfgs, который хорошо справляется с задачами классификации и быстро сходится, если данные соответствующим образом подготовлены.

Малое количество итераций: Несмотря на предупреждение о неудавшейся сходимости при max_iter=200, модель все равно достигла отличных результатов, что говорит о правильной настройке модели для этого типа данных.

**Вывод по второй модели (Mercari Price Suggestion).**

1. Точность модели (Accuracy):
Точность модели составила 70.87%, что ниже, чем в предыдущем примере (97.2%). Это указывает на то, что модель не так хорошо справляется с данной задачей.

2. Матрица ошибок (Confusion Matrix):
Матрица ошибок показывает, что модель часто путает классы, особенно для классов 1 и 3, что видно по большому количеству ошибок на диагонали.
Например, для класса 1 (предположительно более многочисленный) модель ошибается, предсказывая его как класс 0 или 2.
Часть ошибок также наблюдается в классе 3, что может свидетельствовать о сложной классификации этого класса.

3. Отчет о классификации (Classification Report):
Precision, Recall и F1-score для каждого класса демонстрируют хорошие результаты для классов 0 и 1, но значительно хуже для класса 3.
Для класса 3 значения Recall и F1-score значительно ниже (например, 0.43 и 0.53 соответственно), что говорит о сложности правильной классификации данного класса.
Для классов 0 и 1 точность и полнота близки к 1, но для класса 3 модель делает больше ошибок.

Причины таких результатов:

Сложность данных: Возможно, данные для третьего класса имеют особенности, которые затрудняют их различие с другими классами. Это может быть связано с малым количеством образцов или схожестью между классами.

Модель: Логистическая регрессия — это линейная модель, которая может плохо справляться с нелинейными зависимостями в данных. Если данные не имеют явных линейных разделений, модель может давать менее точные результаты.

Пропущенные значения и обработка данных: Хотя в коде предусмотрена обработка пропущенных значений, возможно, что недостаточно хорошая предобработка данных может негативно сказаться на результатах.

**Ответить на вопросы:**
**1. Чем многоклассовая классификация отличается от бинарной?**

Бинарная классификация — это задача с двумя классами, например, "спам/не спам". Модель предсказывает один из двух вариантов.

Многоклассовая классификация — задача с более чем двумя классами, например, "красное/белое/розовое вино". Модель предсказывает один из нескольких классов.

В многоклассовой классификации используются методы типа "один против всех" (One-vs-Rest), а в бинарной — стандартная бинарная модель.

**2. Какие подходы существуют в многоклассовой классификации?**

В многоклассовой классификации существуют несколько подходов:

**One-vs-Rest (OvR) (Один против всех):**

Для каждого класса создается отдельная бинарная модель, которая классифицирует текущий класс как положительный, а остальные — как отрицательные.
Например, если классов 3, то строятся 3 бинарные модели.

**One-vs-One (OvO) (Один против одного):**

Строятся модели для каждой пары классов. Например, для 3 классов будет 3 модели (класс 1 против класса 2, класс 1 против класса 3, класс 2 против класса 3).
Каждая модель решает, какой из двух классов наиболее вероятен.

**Многоклассовая логистическая регрессия:**

Расширение бинарной логистической регрессии, где для каждого класса вычисляется вероятность, и выбирается класс с наибольшей вероятностью.

**Классификация с несколькими выходами (Multilabel classification):**

Это особый случай, когда для одного объекта можно предсказать несколько классов. Например, для одной картинки можно предсказать одновременно несколько объектов.

**Нейронные сети:**

Используют один выходной слой с количеством нейронов, равным числу классов. Сигмоида или софтмакс преобразует вывод в вероятности для каждого класса.
Каждый из этих подходов имеет свои особенности и выбирается в зависимости от задачи.

**3. Какие метрики рассчитываются для многоклассовой классификации?**

**Точность (Accuracy):**

Показывает долю правильно классифицированных примеров среди всех.

**Точность (Precision):**

Оценивает, сколько из предсказанных объектов данного класса действительно принадлежат к этому классу. Это важно, когда важно минимизировать ложные положительные предсказания.

**Полнота (Recall):**

Оценивает, сколько объектов данного класса модель правильно предсказала из всех объектов этого класса. Это важно, когда нужно минимизировать ложные отрицательные предсказания.

**F1-score:**

Это баланс между точностью и полнотой. Чем выше F1-score, тем лучше модель справляется с как точностью, так и полнотой.

**Макро-среднее (Macro-average):**

Среднее значение метрик (точности, полноты, F1) по всем классам, не учитывая количество объектов в классах. Каждый класс имеет равный вес.

**Взвешенное среднее (Weighted-average):**

Среднее значение метрик, учитывающее количество объектов в каждом классе. Классы с большим количеством объектов получают больший вес.

**Матрица ошибок (Confusion Matrix):**

Таблица, которая показывает, как модель классифицировала каждый класс, выделяя правильные и неправильные предсказания для каждого класса.